providers/implementations/signature/rsa_sig.c

```diff
@@ -105,28 +154,196 @@ typedef struct {
+
     /* Temp buffer */
     unsigned char *tbuf;
     OSSL_FIPS_IND_DECLARE
 
 } PROV_RSA_CTX;
 
-/* True if PSS parameters are restricted */
-#define rsa_pss_restricted(prsactx) (prsactx->min_saltlen != -1)
+/*
+ * struct md_wrapper_st functions.  These only support ctx->md.md, not
+ * ctx->mgf1_md, what will demand another set of wrappers.
+ * Some of them look a lot like OSSL_FUNC_digest functions, but are not.
+ */
+static int wrap_md_is_set(PROV_RSA_CTX *ctx)
+{
+    return (ctx->md.md != NULL || ctx->md.init != NULL);
+}
 
-static size_t rsa_get_md_size(const PROV_RSA_CTX *prsactx)
+static int wrap_md_newctx(PROV_RSA_CTX *ctx)
 {
-    int md_size;
+    if (ctx->md.md != NULL)
+        ctx->mdctx = EVP_MD_CTX_new();
+    else if (ctx->md.newctx != NULL)
+        ctx->mdctx = ctx->md.newctx(ctx->provctx);
+    return ctx->mdctx != NULL;
+}
+
+static void wrap_md_freectx(PROV_RSA_CTX *ctx)
+{
+    if (ctx->md.md != NULL)
+        EVP_MD_CTX_free(ctx->mdctx);
+    else if (ctx->md.newctx != NULL)
+        ctx->md.freectx(ctx->mdctx);
+    ctx->mdctx = NULL;
+}
+
+static int wrap_md_init(PROV_RSA_CTX *ctx, const OSSL_PARAM params[])
+{
+    if (ctx->mdctx == NULL)
+        return 0;
+    if (ctx->md.md != NULL)
+        return EVP_DigestInit_ex2(ctx->mdctx, ctx->md.md, params);
+    else if (ctx->md.init != NULL)
+        return ctx->md.init(ctx->mdctx, params);
+    return 0;
+}
+
+static int wrap_md_update(PROV_RSA_CTX *ctx,
+                          const unsigned char *in, size_t inl)
+{
+    if (ctx->mdctx == NULL)
+        return 0;
+    if (ctx->md.md != NULL)
+        return EVP_DigestUpdate(ctx->mdctx, in, inl);
+    else if (ctx->md.update != NULL)
+        return ctx->md.update(ctx->mdctx, in, inl);
+    return 0;
+}
+
+static int wrap_md_final(PROV_RSA_CTX *ctx,
+                         unsigned char *out, size_t *outl, size_t outsz)
+{
+    if (ctx->mdctx == NULL)
+        return 0;
+    if (ctx->md.md != NULL) {
+        unsigned int outlen = (unsigned int)outsz;
+        int ret = EVP_DigestFinal_ex(ctx->mdctx, out, &outlen);
+
+        if (outl != NULL)
+            *outl = (size_t)outlen;
+        return ret;
+    } else if (ctx->md.final != NULL) {
+        return ctx->md.final(ctx->mdctx, out, outl, outsz);
+    }
+    return 0;
+}
+
+static size_t wrap_md_get_size(PROV_RSA_CTX *ctx)
+{
+    if (ctx->md.mdsize > 0)
+        return ctx->md.mdsize;
+
+    if (ctx->md.md != NULL) {
+        int md_size = EVP_MD_get_size(ctx->md.md);
 
-    if (prsactx->md != NULL) {
-        md_size = EVP_MD_get_size(prsactx->md);
         if (md_size <= 0)
             return 0;
         return md_size;
+    } else if (ctx->md.get_params != NULL) {
+        OSSL_PARAM params[2];
+
+        params[0] = OSSL_PARAM_construct_size_t(OSSL_DIGEST_PARAM_SIZE,
+                                                &ctx->md.mdsize);
+        params[1] = OSSL_PARAM_construct_end();
+        if (!ctx->md.get_params(params)) {
+            ctx->md.mdsize = 0;
```

Comment:
IIRC param construction is kind of expensive in terms of cpu cycles, should the mdsize be cached, so this only has to be called once?

Reply:
Did you notice what happens at the very start of this function?